analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/LocksReleasedAllPathsBase.cs

```diff
@@ -179,8 +184,71 @@ private ProgramState RemoveLock(SymbolicContext context, ISymbol symbol)
         private static ProgramState AddLock(SymbolicContext context, IOperation operation) =>
             context.State.SetOperationConstraint(operation, context.SymbolicValueCounter, LockConstraint.Held);
 
-        private static ISymbol FirstArgumentSymbol(IInvocationOperationWrapper invocation) =>
-            IArgumentOperationWrapper.FromOperation(invocation.Arguments.First()).Value.TrackedSymbol();
+        private static ISymbol ArgumentSymbol(IInvocationOperationWrapper invocation, int argumentIndex, string parameterName)
+        {
+            if (IArgumentOperationWrapper.FromOperation(invocation.Arguments[argumentIndex]) is { } argumentOperation
+                && argumentOperation.Parameter.Name == parameterName)
+            {
+                return argumentOperation.Value.TrackedSymbol();
+            }
+            else
+            {
+                for (var i = 0; i < invocation.Arguments.Length; i++)
+                {
+                    if (IArgumentOperationWrapper.FromOperation(invocation.Arguments[i]) is { } namedArgumentOperation
+                        && namedArgumentOperation.Parameter.Name == parameterName)
+                    {
+                        return namedArgumentOperation.Value.TrackedSymbol();
+                    }
+                }
+            }
+            return null;
+        }
+
+        private static RefParamContext FindRefParam(SymbolicContext context) =>
+            InvocationWithLockObjectAndBoolRefParam(context, KnownType.System_Threading_Monitor, 2, 1, "obj", "lockTaken", "Enter", "TryEnter")
+            ?? InvocationWithLockObjectAndBoolRefParam(context, KnownType.System_Threading_Monitor, 3, 2, "obj", "lockTaken", "TryEnter")
+            ?? InvocationWithBoolRefParam(context, KnownType.System_Threading_SpinLock, 1, 0, "lockTaken", "Enter", "TryEnter")
+            ?? InvocationWithBoolRefParam(context, KnownType.System_Threading_SpinLock, 2, 1, "lockTaken", "TryEnter");
+
+        private static RefParamContext InvocationWithLockObjectAndBoolRefParam(SymbolicContext context,
+                                                                               KnownType type,
+                                                                               int parameterCount,
+                                                                               int refParameterIndex,
+                                                                               string lockObjectParameterName,
+                                                                               string refParameterName,
```

Comment:
We should be able to simplify this explicit value. Because we can find the parameter index and name from the target method.
```suggestion
                                                                               string lockObjectParameterName,
```